### PR TITLE
Feature/Extended flip for option weights.

### PR DESCRIPTION
**Pull request by KarlKennerley**
Closes issue #139.

Added the ability to add weights to provided options passed as arguments. The keyword "weights" separates the options and weights in the supplied arguments. The number of options must match the number of weights.

### CODE REVIEW

**Comment by Joeyh021** [comment]
Hey, thanks for the PR! Can you provide a quick screenshot demoing this, and check the CI failures too. Happy to merge otherwise, implementation seems solid.

**Comment by KarlKennerley** [comment]
> Hey, thanks for the PR! Can you provide a quick screenshot demoing this, and check the CI failures too. Happy to merge otherwise, implementation seems solid.

See attached. I've made a further change to the file and will look at the CI now.

Options Weighted
Unmatched Options   Weights

**Comment by ericthelemur** [comment]
Looking good. I would recommend moving "weights" index into a separate variable. That would most likely fix some of Black's formatting problem. The issue it comes up with about remove logging.py is not cause by your code, so ignore that error for now.

**Comment by Joeyh021** [comment]
LGTM

File: cogs/commands/flip.py [hunk 1]
```diff
@@ -22,13 +22,30 @@ async def flip(self, ctx: Context, *, args: str = ""):
...
+                    weighted = True
+                else:
+                    await ctx.send(
+                        f"The number of options and weights supplied must match {display_name}! :confused:"
+                    )
+            else:
+                options = ["Heads", "Tails"] if not args else args
+
+            await ctx.send(
+                f"{display_name}:({random.choices(options, weights=option_weights, k=1)[0] if weighted else {random.choice(options)}.lstrip('@')}"
```
Comment (ericthelemur) [inline cogs/commands/flip.py:47]:
```suggestion
                f"{display_name}: {(random.choices(options, weights=option_weights, k=1)[0] if weighted else {random.choice(options)}).lstrip('@')}"
```
There was a stray bracket after the colon, and moved `lstrip outside the if-else`

File: cogs/commands/flip.py [hunk 2]
```diff
@@ -22,13 +22,30 @@ async def flip(self, ctx: Context, *, args: str = ""):
...
+        else:
+            if "weights" in [arg.lower() for arg in args]:
+                weight_index = args.index("weights")
+                if len(args[:weight_index]) == len(args[: weight_index + 1 :]):
+                    options = args[:weight_index]
+                    option_weights = [
+                        int(weight) for weight in args[weight_index + 1 :]
+                    ]
+                    weighted = True
+                else:
```
Comment (ericthelemur) [inline cogs/commands/flip.py:39]:
```suggestion
                weighted = True
                options = args[:weight_index]
                option_weights = [int(weight) for weight in args[weight_index + 1 :]]
                if len(options) != len(option_weights):
```
I think moving `options` and `options_weights` outside the if statement make the control flow a little clearer. Also there is a stray colon inserted on the RHS of the original condition.

**Review by ericthelemur — CHANGES_REQUESTED** [review]
I've got a couple of suggestions. Moving `options` outside the if makes it a little clearer imo, but everything else is very minor.

(To commit the changes, go to the Files tab, click `Add suggestion to batch` for any you agree with and commit from there)

**Comment by ericthelemur** [comment]
Looks good. I'll give it a final test tonight, then I'll merge. TY

File: cogs/commands/flip.py [hunk 3]
```diff
@@ -22,13 +22,27 @@ async def flip(self, ctx: Context, *, args: str = ""):
...
+                option_weights = [int(weight) for weight in args[weight_index + 1 :]]
+                if len(options) != len(option_weights):
+                    return await ctx.send(
+                        f"The number of options and weights supplied must match {display_name}! :confused:"
+                    )
+            else:
+                options = ["Heads", "Tails"] if not args else args
+
+            await ctx.send(
+                f"{display_name}: {(random.choices(options, weights=option_weights, k=1)[0] if weighted else {random.choice(options)}).lstrip('@')}"
```
Comment (ericthelemur) [inline cogs/commands/flip.py:44]:
```suggestion
                f"{display_name}: {(random.choices(options, weights=option_weights, k=1)[0] if weighted else random.choice(options)).lstrip('@')}"
```

**Review by ericthelemur — APPROVED** [review]

**Pull request merged**